zebra-consensus/src/parameters/network_upgrade.rs

```diff
@@ -0,0 +1,163 @@
+//! Network upgrade consensus parameters for Zcash.
+
+use NetworkUpgrade::*;
+
+use std::collections::{BTreeMap, HashMap};
+use std::ops::Bound::*;
+
+use zebra_chain::types::BlockHeight;
+use zebra_chain::{Network, Network::*};
+
+/// A Zcash network protocol upgrade.
+//
+// TODO: are new network upgrades a breaking change, or should we make this
+//       enum non-exhaustive?
+#[derive(Copy, Clone, Debug, Eq, Hash, PartialEq)]
+pub enum NetworkUpgrade {
+    /// The Zcash protocol before the Overwinter upgrade.
+    ///
+    /// We avoid using `Sprout`, because the specification says that Sprout
+    /// is the name of the pre-Sapling protocol, before and after Overwinter.
+    BeforeOverwinter,
+    /// The Zcash protocol after the Overwinter upgrade.
+    Overwinter,
+    /// The Zcash protocol after the Sapling upgrade.
+    Sapling,
+    /// The Zcash protocol after the Blossom upgrade.
+    Blossom,
+    /// The Zcash protocol after the Heartwood upgrade.
+    Heartwood,
+    /// The Zcash protocol after the Canopy upgrade.
+    Canopy,
+}
+
+/// Mainnet network upgrade activation heights.
+///
+/// This is actually a bijective map, but it is const, so we use a vector, and
+/// do the uniqueness check in the unit tests.
+pub(crate) const MAINNET_ACTIVATION_HEIGHTS: &[(BlockHeight, NetworkUpgrade)] = &[
```

Comment:
How do you feel about splitting these out into associated consts and then creating a single `[(BlockHeight, NetworkUpgrade, ConsensusBranchID)]` for slice iterating over them.

```rust
impl BlockHeight {
    pub const BEFORE_OVERWINTER: BlockHeight = BlockHeight(0);
    ...
}

pub(crate) const NETWORK_UPGRADES: &[(NetworkUpgrade, ConsensusBranchId, BlockHeight)] = &[
    (NetworkUpgrade::BeforeOverwinter, ConsensusBranchID::BEFORE_OVERWINTER, BlockHeight::BEFORE_OVERWINTER),
...
];
```

Reply:
If we decide to do this refactor, I think we should create a "network upgrade parameters" struct, and make them struct instances.

Otherwise, it feels like we're splitting up related items into different impls. (The network upgrade ZIP specifies activation heights and consensus branch IDs in the same section.)

Also, the block height depends on the network (mainnet or testnet), but the consensus branch ID does not. So we'd need to be careful of multiplicity.

Reply:
big agree on the struct thing, not sure how to resolve the multiplicity regarding testnet vs mainnet. Might be worth opening an issue for.